transport.go

```diff
@@ -9,17 +9,63 @@ import (
 	"os"
 	"sync"
 
+	"github.com/gorilla/websocket"
 	"gopkg.in/square/go-jose.v2/json"
 )
 
-type transport struct {
+type Transport interface {
+	Start() error
+	Stop() error
+	Send(message map[string]interface{}) error
+	SetDispatch(dispatch func(msg *message))
+}
+
+type pipeTransport struct {
 	stdin    io.WriteCloser
 	stdout   io.ReadCloser
 	dispatch func(msg *message)
 	rLock    sync.Mutex
 }
 
-func (t *transport) Start() error {
+type websocketTransport struct {
+	url      string
+	conn     *websocket.Conn
+	dispatch func(msg *message)
+}
+
+func (t *websocketTransport) Start() error {
+	conn, _, err := websocket.DefaultDialer.Dial(t.url, nil)
+	if err != nil {
+		return fmt.Errorf("could not connect to websocket: %w", err)
+	}
+	t.conn = conn
+	for {
+		msg := &message{}
+		err := t.conn.ReadJSON(msg)
```

Comment:
Single error returns get usually handled inside the if condition. It's shorter. 

```suggestion
		if err := t.conn.ReadJSON(msg); err != nil {
       ...
```

Reply:
Next time